controllers/addon_controller.go

```diff
@@ -173,71 +172,77 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
 	return ret, procErr
 }
 
-// SetupWithManager is called to setup manager and watchers
-func (r *AddonReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	log := r.Log
-	managedNS := "addon-manager-system"
-
-	nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, managedNS, nil)
-	wfInf := nsInformers.ForResource(common.WorkflowGVR())
-	bldr := ctrl.NewControllerManagedBy(mgr).
-		For(&addonmgrv1alpha1.Addon{}).
-		// Watch workflows created by addon only in addon-manager-system namespace
-		Watches(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
-			IsController: true,
-			OwnerType:    &addonmgrv1alpha1.Addon{},
-		})
-
-	generatedInformers = informers.NewSharedInformerFactory(r.generatedClient, time.Minute*30)
-
-	err := mgr.Add(manager.RunnableFunc(func(s <-chan struct{}) error {
-		generatedInformers.Start(s)
-		generatedInformers.WaitForCacheSync(s)
-		nsInformers.Start(s)
-		nsInformers.WaitForCacheSync(s)
-		<-s
-		return nil
-	}))
+func New(mgr manager.Manager) (controller.Controller, error) {
+	r := &AddonReconciler{
+		Client:          mgr.GetClient(),
+		Log:             ctrl.Log.WithName(controllerName),
+		Scheme:          mgr.GetScheme(),
+		versionCache:    addon.NewAddonVersionCacheClient(),
+		dynClient:       dynamic.NewForConfigOrDie(mgr.GetConfig()),
+		generatedClient: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		recorder:        mgr.GetEventRecorderFor("addons"),
+		statusWGMap:     map[string]*sync.WaitGroup{},
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		log.Error(err, "Error adding InformerFactory to the Manager")
-		return err
+		return nil, err
+	}
+
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Kind{Type: &wfv1.Workflow{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}); err != nil {
+		return nil, err
 	}
 
 	// Watch for changes to kubernetes Resources matching addon labels.
-	for _, resc := range resources {
-		gvk := resc.GetObjectKind().GroupVersionKind()
-		_, kind := gvk.ToAPIVersionAndKind()
-
-		gvr := schema.GroupVersionResource{
-			Group:    gvk.Group,
-			Version:  gvk.Version,
-			Resource: inflection.Plural(strings.ToLower(kind)),
-		}
+	if err := c.Watch(&source.Kind{Type: &addonmgrv1alpha1.Addon{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
 
-		inf, err := generatedInformers.ForResource(gvr)
-		if err != nil {
-			return err
-		}
+	if err := c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, r.enqueRequestWithAddonLabel()); err != nil {
```

Comment:
We had a for loop of `resources` before with same types, why we move to single mention; this makes the [resources array](https://github.com/keikoproj/addon-manager/pull/95/files#diff-2df6a56a33260adb76f704ff7c6737be9f43c13fe523521ee76390ff2a9ca4b8R60) unused.

Reply:
The resource leverage a SharedInformers which is widely used in previous version. Newer Version, we can leverage the client, watch/retrieve/filter resource directly, cause we have a generic handler for generic kubernetes handler